pkg/image/controller/trigger/cache.go

```diff
@@ -0,0 +1,128 @@
+package trigger
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/client/cache"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	"github.com/openshift/origin/pkg/image/api/v1/trigger"
+)
+
+type TriggerFunc func(interface{}) (*TriggerCacheEntry, bool, error)
+
+func AnnotationTrigger(obj interface{}) (*TriggerCacheEntry, bool, error) {
+	m, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, false, err
+	}
+	t, ok := m.GetAnnotations()["image.openshift.io/triggers"]
+	if !ok {
+		return nil, false, nil
+	}
+	r := &TriggerCacheEntry{}
+	if err := json.Unmarshal([]byte(t), &r.Triggers); err != nil {
+		return nil, false, err
+	}
+	if len(r.Triggers) == 0 {
+		return nil, false, nil
+	}
+	r.Namespace = m.GetNamespace()
+	r.Key = r.Namespace + "/" + m.GetName()
+	return r, true, nil
+}
+
+type triggerCache struct {
+	cache.ThreadSafeStore
+}
+
+type TriggerCacheEntry struct {
+	Key       string
+	Namespace string
+	Triggers  []trigger.ObjectFieldTrigger
+}
+
+func newTriggerCache() triggerCache {
+	return triggerCache{
+		ThreadSafeStore: cache.NewThreadSafeStore(
+			cache.Indexers{
+				"images": func(obj interface{}) ([]string, error) {
+					entry := obj.(*TriggerCacheEntry)
+					var keys []string
+					for _, t := range entry.Triggers {
+						if t.From.Kind != "ImageStreamTag" || len(t.From.APIVersion) != 0 || t.Paused {
```

Comment:
@smarterclayton this will work just for images or do we plan to make it generic enough to work with other resource types, like secrets or config maps?